### PR TITLE
Add CMakeLists.txt generated with scram2cmake.py

### PR DESCRIPTION
**Pull request by gartung**

### CODE REVIEW

**Comment by gartung** [comment]
@Dr15Jones

**Comment by Dr15Jones** [comment]
please test

**Comment by davidlange6** [comment]
to avoid developer confusion - could the cmakelists.txt files be autogenerated instead? @smuzaffar 

(and we should find a better home for cmaketools and scram2make)

**Comment by davidlange6** [comment]
autogeneration would also have the advantage of keeping things up to date without humans doing work..

**Comment by gartung** [comment]
There are some corner cases where BuildFile.xml is nor interpreted as XML correctly.

**Comment by gartung** [comment]
@sextonkennedy   requested Stitched be made buildable and eventually moved to the HSF orginization. Since there is no way to build Stiched independently of scram at the moment I think adding the generated CMakeLists.txt is a good solution

**Comment by davidlange6** [comment]
can we fix those? Whats an example?

File: Utilities/General/test/BuildFile.xml [hunk 1]
```diff
@@ -9,8 +9,3 @@
   <use name="google-benchmark-main"/>
 </bin>
 
-<iftool name="cuda-gcc-support">
```
Comment (gartung) [inline Utilities/General/test/BuildFile.xml]: @davidlange6 This is not handled by scram2cmake.py or scram2cmake.pl

File: IOMC/RandomEngine/test/BuildFile.xml [hunk 2]
```diff
@@ -7,7 +7,9 @@
 <use   name="clhep"/>
 <library   file="TestRandomNumberServiceGlobal.cc" name="IOMCRandomEngineTestRandomNumberServiceGlobal">
   <flags   EDM_PLUGIN="1"/>
+  <use   name="clhep"/>
```
Comment (gartung) [inline IOMC/RandomEngine/test/BuildFile.xml]: Here I needed to propagate the library dependency.

File: IOMC/RandomEngine/test/BuildFile.xml [hunk 3]
```diff
@@ -7,7 +7,9 @@
...
 </library>
 <bin   file="TestIOMCRandomEngineService.cpp">
   <flags   TEST_RUNNER_ARGS=" /bin/bash IOMC/RandomEngine/test testRandomService.sh"/>
+  <use   name="clhep"/>
```
Comment (gartung) [inline IOMC/RandomEngine/test/BuildFile.xml]: Propagate library dependency

File: DataFormats/FWLite/test/BuildFile.xml [hunk 4]
```diff
@@ -4,15 +4,19 @@
 <bin   name="testDataFormatsFWLiteMakeInput" file="TestRunnerDataFormatsFWLite.cpp">
   <flags   TEST_RUNNER_ARGS=" /bin/bash DataFormats/FWLite/test RefTest.sh"/>
   <use   name="FWCore/Utilities"/>
+  <use   name="cppunit"/>
```
Comment (gartung) [inline DataFormats/FWLite/test/BuildFile.xml]: Propagate library dependency

File: DataFormats/FWLite/test/BuildFile.xml [hunk 5]
```diff
@@ -4,15 +4,19 @@
...
+  <use   name="boost_filesystem"/>
```
Comment (gartung) [inline DataFormats/FWLite/test/BuildFile.xml]: same

File: DataFormats/FWLite/test/BuildFile.xml [hunk 6]
```diff
@@ -4,15 +4,19 @@
...
 </bin>
 <bin   name="testDataFormatsFWLite" file="test.cppunit.cpp,format_type_name.cppunit.cpp">
   <flags PRE_TEST="testDataFormatsFWLiteMakeInput"/>
   <use   name="DataFormats/FWLite"/>
   <use   name="FWCore/FWLite"/>
   <use   name="DataFormats/TestObjects"/>
+  <use   name="cppunit"/>
```
Comment (gartung) [inline DataFormats/FWLite/test/BuildFile.xml]: same

File: DataFormats/Provenance/test/BuildFile.xml [hunk 7]
```diff
@@ -8,6 +8,7 @@
   <use name="catch2"/>
 </bin>
 <bin   name="testProductResolverIndexHelper" file="testRunner.cpp productResolverIndexHelper_t.cppunit.cc">
+<use   name="cppunit"/>
```
Comment (gartung) [inline DataFormats/Provenance/test/BuildFile.xml]: same

File: DataFormats/TestObjects/test/BuildFile.xml [hunk 8]
```diff
@@ -1,6 +1,7 @@
 <use   name="cppunit"/>
 <use   name="DataFormats/TestObjects"/>
 <bin   file="Enum_t.cpp">
+  <use   name="cppunit"/>
```
Comment (gartung) [inline DataFormats/TestObjects/test/BuildFile.xml]: Propagate library dependency

File: FWCore/FWLite/bin/BuildFile.xml [hunk 9]
```diff
@@ -2,12 +2,14 @@
   <use   name="rootcore"/>
   <use   name="boost"/>
   <use   name="boost_program_options"/>
+  <use   name="tinyxml2"/>
```
Comment (gartung) [inline FWCore/FWLite/bin/BuildFile.xml]: Propagate library dependency

File: FWCore/FWLite/bin/BuildFile.xml [hunk 10]
```diff
@@ -2,12 +2,14 @@
...
   <use   name="FWCore/FWLite"/>
 </bin>
 
 <bin   name="edmBranchStorageSize" file="storageSizeForBranch.cc">
   <use   name="rootcore"/>
   <use   name="boost"/>
   <use   name="boost_program_options"/>
+  <use   name="tinyxml2"/>
```
Comment (gartung) [inline FWCore/FWLite/bin/BuildFile.xml]: same

File: FWCore/FWLite/test/BuildFile.xml [hunk 11]
```diff
@@ -4,8 +4,10 @@
 <bin   name="testFWCoreFWLite" file="ref_t.cppunit.cpp">
   <use   name="FWCore/FWLite"/>
   <use   name="DataFormats/TestObjects"/>
+  <use   name="cppunit"/>
```
Comment (gartung) [inline FWCore/FWLite/test/BuildFile.xml]: Propagate library dependency

File: FWCore/FWLite/test/BuildFile.xml [hunk 12]
```diff
@@ -4,8 +4,10 @@
...
 </bin>
 <bin   file="TestRunnerFWCoreFWLite.cpp">
   <flags   TEST_RUNNER_ARGS=" /bin/bash FWCore/FWLite/test run_all_t.sh"/>
   <use   name="FWCore/Utilities"/>
+  <use   name="FWCore/TFWLiteSelector"/>
```
Comment (gartung) [inline FWCore/FWLite/test/BuildFile.xml]: same

File: FWCore/ServiceRegistry/test/BuildFile.xml [hunk 13]
```diff
@@ -1,21 +1,30 @@
+<library   file="stubs/DummyService.cc,stubs/DummyServiceE0.cc,stubs/DependsOnDummyService.cc,stubs/DummyStoreConfigService.cc" name="FWCoreServiceRegistryTestDummyService">
 <use   name="boost"/>
 <use   name="cppunit"/>
 <use   name="FWCore/ServiceRegistry"/>
-<library   file="stubs/DummyService.cc,stubs/DummyServiceE0.cc,stubs/DependsOnDummyService.cc,stubs/DummyStoreConfigService.cc" name="FWCoreServiceRegistryTestDummyService">
 </library>
 <library   file="stubs/Modules.cc" name="FWCoreServiceRegistryTestDummyServicePlugin">
   <lib   name="FWCoreServiceRegistryTestDummyService"/>
+<use   name="boost"/>
```
Comment (gartung) [inline FWCore/ServiceRegistry/test/BuildFile.xml]: Propagate library dependencies

File: FWCore/ServiceRegistry/test/BuildFile.xml [hunk 14]
```diff
@@ -1,21 +1,30 @@
...
+<use   name="cppunit"/>
+<use   name="FWCore/ServiceRegistry"/>
   <flags   EDM_PLUGIN="1"/>
 </library>
 <bin   name="testServiceRegistry" file="serviceregistry_t.cppunit.cpp,servicesmanager_t.cppunit.cc,connect_but_block_self_t.cppunit.cc">
   <lib   name="FWCoreServiceRegistryTestDummyService"/>
   <use   name="FWCore/ParameterSet"/>
   <use   name="FWCore/PluginManager"/>
   <use   name="FWCore/Utilities"/>
+<use   name="boost"/>
```
Comment (gartung) [inline FWCore/ServiceRegistry/test/BuildFile.xml]: same

File: FWCore/ServiceRegistry/test/BuildFile.xml [hunk 15]
```diff
@@ -1,21 +1,30 @@
...
+<use   name="cppunit"/>
+<use   name="FWCore/ServiceRegistry"/>
 </bin>
 <bin   name="servicesmanager_order" file="servicesmanager_order.cpp">
   <lib   name="FWCoreServiceRegistryTestDummyService"/>
   <use   name="FWCore/ParameterSet"/>
   <use   name="FWCore/PluginManager"/>
   <use   name="FWCore/Utilities"/>
+<use   name="boost"/>
```
Comment (gartung) [inline FWCore/ServiceRegistry/test/BuildFile.xml]: same

File: Utilities/General/test/BuildFile.xml (hunk 1 above)
Comment (davidlange6) [reply Utilities/General/test/BuildFile.xml]: its valid xml, so I guess scram2cmake needs to be told about it  - is that hard?

**Comment by smuzaffar** [comment]
I agree we should fix the build files if there are any missing dependencies or if the format is wrong.

File: Utilities/General/test/BuildFile.xml (hunk 1 above)
Comment (smuzaffar) [reply Utilities/General/test/BuildFile.xml]: scram2cmake.py should not see flags, it makes use of SCRAM caches. You might have an old version of SCRAM which does not know about this flag.

Comment (gartung) [reply Utilities/General/test/BuildFile.xml]: The error is 
<code>
error in  Utilities/General/test/BuildFile.xml: not well-formed (invalid token): line 15, column 78
</code>
 This is the &&.

Comment (gartung) [reply Utilities/General/test/BuildFile.xml]: Changing to  &amp;&amp; removes the error. I will make a pull request with the BuildFile.xml changes for testing.

Comment (gartung) [reply Utilities/General/test/BuildFile.xml]: @smuzaffar note I am using scram2cmake.PY so scram is never invoked.

Comment (gartung) [reply Utilities/General/test/BuildFile.xml]: A better name for the tool would be buildfile2cmake.py

**Comment by Dr15Jones** [comment]
@gartung please also make a pull request to CMSSW with the same fix.

File: DataFormats/Provenance/test/BuildFile.xml (hunk 7 above)
Comment (davidlange6) [reply DataFormats/Provenance/test/BuildFile.xml]: is line 1 not meant to do that already? (eg difference of xml interpretation between scram and the python tool?)

Comment (gartung) [reply DataFormats/Provenance/test/BuildFile.xml]: Yes but I never figured out how. I was modifying the origin which did not consider binaries to suit my needs.https://github.com/Teemperor/scram2cmake

File: Utilities/General/test/BuildFile.xml (hunk 1 above)
Comment (smuzaffar) [reply Utilities/General/test/BuildFile.xml]: better to change it to ; which will allow scram to build it too.

**Comment by smuzaffar** [comment]
@gartung we should not change any thing in this repository otherwise we will not be able to push changes from cmssw.

**Comment by smuzaffar** [comment]
`&&` can be replaced by `;` in cmssw . This like is just a shell command run by `scram build runtests`. So it use `;` should allow it to work with scram

**Comment by davidlange6** [comment]
i gather this the code that should do it

https://github.com/gartung/scram2cmake/blob/master/scram2cmake.py#L603

**Comment by gartung** [comment]
@smuzaffar I will try the replacing && with ; and running the test with scram to make sure it works. After that I will submit a pull request to CMSSW.

I intend to close this pull request without merging once I can get the CMakeLists.txt generation fully automated.

I worked on this two years ago and it went nowhere because I never make if "official".

**Comment by gartung** [comment]
If we add an "official" buildfile2cmake.py tool to the CMS-SW organization it could be used with and FWLite repo as well. I would like to make the FWLite conda build "official".

**Comment by gartung** [comment]
Closing for now to quite the update emails.

**Pull request closed without merging**